File: 백준/Silver/10451. 순열 사이클/순열 사이클.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(2000)

def cycle(nums, chk, idx, c):
    if chk[idx] != 0:
        return 1 if chk[idx] == c else 0
    chk[idx] = c
    return cycle(nums, chk, nums[idx] - 1, c)

def solve():
    T = int(input())
    for _ in range(T):
        N = int(input())
        nums = [*map(int, input().split())]
        chk = [0] * N
        ans = 0
        for i in range(N):
            if chk[i] == 0:
                ans += cycle(nums, chk, i, i + 1)
        print(ans)

solve()

File: 백준/Silver/10451. 순열 사이클/순열 사이클.java
import java.io.*;
import java.util.*;

public class Main {

    static int readInt(String str) { return Integer.parseInt(str); }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int T = readInt(br.readLine());
        for (int tc = 0; tc < T; tc++) {
            int n = readInt(br.readLine());
            st = new StringTokenizer(br.readLine());
            int[] num = new int[n];
            for (int i = 0; i < n; i++) {
                num[i] = readInt(st.nextToken());
            }
            int ans = 0;
            int[] chk = new int[n];
            for (int i = 0; i < n; i++) {
                if (chk[i] == 0) {
                    ans += cycle(num, chk, i, i + 1);
                }
            }
            sb.append(ans).append("\n");
        }
        System.out.println(sb);
    }

    static int cycle(int[] num, int[] chk, int idx, int c) {
        if (chk[idx] != 0) {
            return chk[idx] == c ? 1 : 0;
        }
        chk[idx] = c;
        return cycle(num, chk, num[idx] - 1, c);
    }

}